### ISSUE TITLE
Speedup of cuda histogram - sanity check

### ISSUE THREAD

**Issue by anshumang** [arrayfire/arrayfire#1313]
With the following modified `histogramKernel` -
`template<typename inType, typename outType, bool isLinear>
static __global__
void histogramKernel(Param<outType> out, CParam<inType> in, unsigned long long int *d_nb_blocks_completed, unsigned long long int tot_nb_blocks,
                     int len, int nbins, float minval, float maxval, int nBBS)
{
    __shared__ unsigned long long int s_nb_blocks_completed;
    SharedMemory<outType> shared;
    outType * shrdMem = shared.getPointer();

    while(true){
            
    if(threadIdx.x==0){
            s_nb_blocks_completed = atomicAdd(d_nb_blocks_completed, 1); 
    }   
    __syncthreads();
    if(s_nb_blocks_completed>=tot_nb_blocks){
            return;
    }   

    // offset input and output to account for batch ops
    //unsigned b2 = blockIdx.x / nBBS;
    unsigned b2 = s_nb_blocks_completed / nBBS;
    const inType *iptr  =  in.ptr + b2 *  in.strides[2] + blockIdx.y *  in.strides[3];
    outType      *optr  = out.ptr + b2 * out.strides[2] + blockIdx.y * out.strides[3];

    //int start  = (blockIdx.x-b2*nBBS) * THRD_LOAD * blockDim.x + threadIdx.x;
    int start  = (s_nb_blocks_completed-b2*nBBS) * THRD_LOAD * blockDim.x + threadIdx.x;
    int end    = minimum((start + THRD_LOAD * blockDim.x), len);
    float step = (maxval-minval) / (float)nbins;

    for (int i = threadIdx.x; i < nbins; i += blockDim.x)
        shrdMem[i] = 0;
    __syncthreads();

    for (int row = start; row < end; row += blockDim.x) {
        int idx = isLinear ? row : ((row % in.dims[0]) + (row / in.dims[0])*in.strides[1]);
        int bin = (int)((iptr[idx] - minval) / step);
        bin     = (bin < 0)      ? 0         : bin;
        bin     = (bin >= nbins) ? (nbins-1) : bin;
        atomicAdd((shrdMem + bin), 1); 
    }   
    __syncthreads();

    for (int i = threadIdx.x; i < nbins; i += blockDim.x) {
        atomicAdd((optr + i), shrdMem[i]);
    }   

    }//end while(true)
}`
and the `histogram` launcher function-  
`bool g_init=false;
unsigned long long int *d_nb_blocks_completed=NULL;

template<typename inType, typename outType, bool isLinear>
void histogram(Param<outType> out, CParam<inType> in, int nbins, float minval, float maxval)
{
    dim3 threads(kernel::THREADS_X, 1);

    int nElems = in.dims[0] * in.dims[1];
    int blk_x  = divup(nElems, THRD_LOAD*THREADS_X);

    dim3 blocks(blk_x * in.dims[2], in.dims[3]);

    int smem_size = nbins * sizeof(outType);

    //CUDA_LAUNCH_SMEM((histogramKernel<inType, outType, isLinear>), blocks, threads, smem_size,
    //        out, in, nElems, nbins, minval, maxval, blk_x);
    unsigned long long int tot_nb_blocks=blocks.x, tmp=0;
    if(!g_init){
            cudaMalloc(&d_nb_blocks_completed, sizeof(unsigned long long int));
            g_init=true;
    }
    cudaMemcpy(d_nb_blocks_completed, &tmp, sizeof(unsigned long long int), cudaMemcpyHostToDevice);
    histogramKernel<inType, outType, isLinear><<<90, threads, smem_size>>>(out, in, d_nb_blocks_completed, tot_nb_blocks, nElems, nbins, minval, maxval, blk_x);

    POST_LAUNCH_CHECK();
}`
I notice 6x speedup on a K40 for an input of 128M double-precision numbers (1GB).

In summary, I only launch 6 threads blocks per SM (x 15 SMs on K40) where 6 is the maximum # of thread blocks that fit on a K40 SM for `histogramKernel`. Each of these 90 thread blocks iterates in a loop and atomically increments a counter in the global memory until the counter reaches the original grid size (16K for the above input size). The other change is replacing blockIdx.x with the observed value of the counter.

Any thoughts?

**Comment by pavanky** [comment]
@anshumang The performance will not only vary with size but also with the data distribution because of atomic operations. 

That said, can you send this in a PR format to `devel` branch ? That way we'd be able to review it by looking at the differences.

**Comment by pavanky** [comment]
@anshumang 
Actually reading through the kernel code it looks like a solid solution.

Only suggestion is to not hard code the blocks and threads. Instead query the CUDA device properties and launch the appropriate blocks threads from there.

An easy way to get the current device properties is to use
 `cudaDeviceProp prop = getDeviceProp(getActiveDeviceId())`.

A PR with this change will be greatly appreciated :+1:

**Comment by pavanky** [comment]
@anshumang you know what, send the PR to `hotfix-3.3.1` instead since this is not changing the API. Make sure the tests are passing.

**Comment by pavanky** [comment]
Can you also move this to the outside of the while loop and see if this improves performance ?
```
  for (int i = threadIdx.x; i < nbins; i += blockDim.x) {
        atomicAdd((optr + i), shrdMem[i]);
    }   
```
This change means you'll have to `break` from while instead of calling `return`.

**Comment by anshumang** [comment]
Thanks @pavanky for the review and feedback. Getting back with the PR with review comments addressed. Is it ok to only address 1D blocks in this PR?

**Comment by pavanky** [comment]
@anshumang You can keep `blocks.y` the same as it was before. It should not drop the performance for 1D  and 2D inputs.
